File: benchmarks/pandas/bench_merge_asof.py
"""Benchmark: merge_asof — backward asof join of two 10k-row sorted DataFrames"""
import json
import time
import pandas as pd

N = 10_000
WARMUP = 3
ITERATIONS = 10

# Trades sorted by time: 0, 2, 4, ...
trade_times = list(range(0, N * 2, 2))
prices = [100.0 + i * 0.5 for i in range(N)]

# Quotes sorted by time, sparser: 0, 3, 6, ...
quote_times = list(range(0, N * 3, 3))
bids = [99.0 + i * 0.5 for i in range(N)]

trades = pd.DataFrame({"time": trade_times, "price": prices})
quotes = pd.DataFrame({"time": quote_times, "bid": bids})

for _ in range(WARMUP):
    pd.merge_asof(trades, quotes, on="time")

start = time.perf_counter()
for _ in range(ITERATIONS):
    pd.merge_asof(trades, quotes, on="time")
total = (time.perf_counter() - start) * 1000

print(json.dumps({
    "function": "merge_asof",
    "mean_ms": total / ITERATIONS,
    "iterations": ITERATIONS,
    "total_ms": total,
}))

File: benchmarks/tsb/bench_merge_asof.ts
/**
 * Benchmark: mergeAsof — backward asof join of two 10k-row sorted DataFrames.
 * Outputs JSON: {"function": "merge_asof", "mean_ms": ..., "iterations": ..., "total_ms": ...}
 */
import { DataFrame, mergeAsof } from "../../src/index.ts";

const N = 10_000;
const WARMUP = 3;
const ITERATIONS = 10;

// Trades sorted by time: 0, 2, 4, ...
const tradeTimes = Array.from({ length: N }, (_, i) => i * 2);
const prices = Array.from({ length: N }, (_, i) => 100.0 + i * 0.5);

// Quotes sorted by time, sparser: 0, 3, 6, ...
const quoteTimes = Array.from({ length: N }, (_, i) => i * 3);
const bids = Array.from({ length: N }, (_, i) => 99.0 + i * 0.5);

const trades = DataFrame.fromColumns({ time: tradeTimes, price: prices });
const quotes = DataFrame.fromColumns({ time: quoteTimes, bid: bids });

for (let i = 0; i < WARMUP; i++) {
  mergeAsof(trades, quotes, { on: "time" });
}

const start = performance.now();
for (let i = 0; i < ITERATIONS; i++) {
  mergeAsof(trades, quotes, { on: "time" });
}
const total = performance.now() - start;

console.log(
  JSON.stringify({
    function: "merge_asof",
    mean_ms: total / ITERATIONS,
    iterations: ITERATIONS,
    total_ms: total,
  }),
);
